File: 2021/clojure/day0401.clj
(comment
  "Author: Loading BG#7962")

(defn check [nums]
  (loop [i 0]
    (cond
      (= i (dec (count nums))) "increasing"
      (> (get nums i) (get nums (inc i)))
        (loop [j 0]
          (cond
            (= j (dec (count nums))) "decreasing"
            (< (get nums j) (get nums (inc j))) "neither"
            :else (recur (inc j))))
      :else (recur (inc i)))))

(defn within-tri? [[[x1 y1] [x2 y2] [x3 y3]] [x y]]
  (.contains (java.awt.Polygon. (int-array [x1 x2 x3]) (int-array [y1 y2 y3]) 3) x y))

(defn color [pos]
  (= (mod (int (get pos 0)) 2) (mod (int (get pos 1)) 2)))
(defn bishop [start end moves]
  (= (color start) (color end)))

(println "check #1:" (check [1 2 3]))
(println "check #2:" (check [3 2 1]))
(println "check #3:" (check [1 2 1]))
(println "---------")
(println "withinTriangle #1:" (within-tri? ['(1 4) '(5 6) '(6 1)] '(4 5)))
(println "withinTriangle #2:" (within-tri? ['(1 4) '(5 6) '(6 1)] '(3 2)))
(println "---------")
(println "bishop #1:" (bishop "a1" "b4" 2))
(println "bishop #2:" (bishop "a1" "b5" 5))
(println "bishop #3:" (bishop "f1" "f1" 0))

File: 2021/kotlin/Day0401.kt
/**
 * Author : Zone-Infinity#9999
 */

import java.awt.Polygon

fun main() {
    println(check(listOf(1, 2, 3)))
    println(check(listOf(3, 2, 1)))
    println(check(listOf(1, 2, 1)))
    println("-----")
    println(withinTriangle(intArrayOf(1, 4), intArrayOf(5, 6), intArrayOf(6, 1), intArrayOf(4, 5)))
    println(withinTriangle(intArrayOf(1, 4), intArrayOf(5, 6), intArrayOf(6, 1), intArrayOf(3, 2)))
    println("-----")
    println(bishop("a1", "b4", 2))
    println(bishop("a1", "b5", 5))
    println(bishop("f1", "f1", 0))
}

fun check(i: List<Int>): String {
    if (i == i.sorted()) return "increasing"
    if (i == i.sorted().reversed()) return "decreasing"
    return "neither"
}

fun withinTriangle(a: IntArray, b: IntArray, c: IntArray, testPoint: IntArray) =
        Polygon(intArrayOf(a[0], b[0], c[0]), intArrayOf(a[1], b[1], c[1]), 3).contains(testPoint[0], testPoint[1])

fun color(s: String) = s[0].toInt() % 2 == s[1].toInt() % 2
fun bishop(a: String, b: String, moves: Int) = color(a) == color(b)